File: scripts/clap_schema_parity.py
#!/usr/bin/env python3
"""Clap-vs-schema parity: every flag clap accepts must appear in `schema --cmd`.

WHY THIS EXISTS AS A SECOND GATE
  `schema-drift-check.sh` compares `docs/schemas/*.json` against the live binary
  and is correct on its own axis. It cannot see this failure, because both sides
  of that comparison derive from the SAME hand-written schema module: the
  generator asks the binary, the binary asks `src/commands/meta/schema/*.rs`.
  When a flag is added to the clap enum and wired in the dispatcher, nothing in
  that loop forces a schema edit, so the file and the binary agree perfectly
  about a surface neither of them describes.

  An agent discovers what it may pass by reading `schema`. A flag absent there
  does not exist for the agent, whatever the parser happens to accept.

WHAT THIS GATE FIRST REPORTED, AND WHY IT WAS WRONG
  The first cut reported 29 undocumented flags across `console`, `heap`, `mitm`,
  `net`, `perf` and `storage`, and this docstring asserted that
  `schema --cmd console` published only `action`, `id` and `path`.

  Re-measured 2026-08-06 against the same binary: the top level does carry only
  those three, but `--page-idx`, `--page-size`, `--types`, `--include-preserved`
  and `--service-worker-id` are all published under
  `action.actions.list.properties`. The schema described them the whole time.

  Every one of the 29 was a false positive with a single cause: the clap side
  unions `<cmd> --help` with `<cmd> <sub> --help`, so it sees two levels, while
  the schema side read one. The six failing commands are exactly the six that
  dispatch on an action word -- which is the signature of a depth bug, not of a
  documentation gap. See `_collect_flags`.

  Recorded rather than deleted, because the failure mode matters more than the
  finding: a gate is a measuring instrument, and an instrument that disagrees
  with the artifact is not evidence about the artifact until the instrument
  itself has been checked.

WHY `--help` AND NOT A PARSE OF THE RUST SOURCE
  The same reason the drift gate trusts the binary: `--help` is what clap
  actually built, so it cannot disagree with the parser. A regex over
  `src/cli/` would introduce a third opinion, and a third opinion is a third
  thing that can be wrong on its own.
"""

from __future__ import annotations

import json
import re
import subprocess
import sys

# A flag line in clap help: two or more leading spaces, an optional short alias,
# then the long name. `--flag[=<V>]`, `--flag <V>` and bare `--flag` all match.
FLAG_RE = re.compile(r"^\s{2,}(?:-[a-zA-Z], )?(--[a-z0-9][a-z0-9-]*)")

# A subcommand line inside the `Commands:` block: two spaces, the name, then at
# least two spaces before its description.
SUBCMD_RE = re.compile(r"^\s{2}([a-z][a-z0-9-]*)\s{2,}\S")

# Sections that follow the command's own `Options:` and carry global flags only.
# Kept for documentation value; the global set is subtracted wholesale below, so
# correctness does not depend on this list being complete.
GLOBAL_SECTIONS = ("Output:", "Timeouts:", "Capabilities:", "Network:")


def run(binary: str, args: list[str]) -> str:
    """Capture stdout+stderr of a help invocation; clap may use either."""
    proc = subprocess.run(
        [binary, *args],
        capture_output=True,
        text=True,
        check=False,
    )
    return proc.stdout + proc.stderr


def flags_in(text: str) -> set[str]:
    """Every long flag clap prints for this invocation."""
    return {m.group(1) for line in text.splitlines() if (m := FLAG_RE.match(line))}


def subcommands_in(text: str) -> set[str]:
    """Names listed under the `Commands:` block, if the command has one."""
    out: set[str] = set()
    in_block = False
    for line in text.splitlines():
        if line.startswith("Commands:"):
            in_block = True
            continue
        if in_block:
            # A non-indented line ends the block (`Options:`, `Arguments:`, ...).
            if line and not line.startswith(" "):
                break
            if m := SUBCMD_RE.match(line):
                out.add(m.group(1))
    return out - {"help"}


def _collect_flags(props: dict, out: set[str]) -> None:
    """Add every long flag a properties object declares, at any nesting depth.

    WHY THE WALK RECURSES
      A command that dispatches on an action word keeps each action's arguments
      under `action.actions.<name>.properties`, not at the top level. Reading
      only the top level while the clap side already unions `<cmd> <sub> --help`
      compares two different depths, and every nested flag is then reported as
      undocumented.

      Measured 2026-08-06: that asymmetry alone produced all 29 findings of this
      gate's first cut. `storage export --path` was cited as unreachable by an
      agent while the schema published it at
      `action.actions.export.properties.path`; `console list --page-idx` was
      cited the same way and sits at `action.actions.list.properties.page_idx`.

      This is the fictional-debt failure named in the module docstring, repeated
      one level down: an instrument that reads the wrong depth does not merely
      miss real debt, it invents debt that costs real work to disprove. The
      first cut of this gate paid that price on the `argv` field; this is the
      same lesson on the nesting axis.
    """
    for key, prop in props.items():
        if not isinstance(prop, dict):
            out.add("--" + key.replace("_", "-"))
            continue
        argv = prop.get("argv", "")
        if isinstance(argv, str) and argv.startswith("--"):
            out.add(argv)
        else:
            out.add("--" + key.replace("_", "-"))
        actions = prop.get("actions")
        if isinstance(actions, dict):
            for spec in actions.values():
                if isinstance(spec, dict):
                    nested = spec.get("properties")
                    if isinstance(nested, dict):
                        _collect_flags(nested, out)


def schema_flags(binary: str, command: str) -> set[str] | None:
    """Long flags the published schema declares for `command`.

    Returns None when the command has no schema at all, which is a different
    failure from a schema that omits flags and is reported separately.

    WHY THE KEY IS A FALLBACK FOR `argv`
      Only some properties carry an explicit `argv`; most identify themselves by
      key alone. A first cut of this gate collected `argv` and nothing else, and
      reported 127 undocumented flags -- including `qr --text`, which the schema
      documents as the property `text`. That is the fictional-debt failure this
      repository already paid for once: an instrument that measures the wrong
      field does not merely miss real debt, it invents debt that costs real work
      to disprove. The property key IS the documentation when `argv` is absent,
      so it counts.
    """
    proc = subprocess.run(
        [binary, "--json", "schema", "--cmd", command],
        capture_output=True,
        text=True,
        check=False,
    )
    if proc.returncode != 0:
        return None
    try:
        payload = json.loads(proc.stdout)
    except json.JSONDecodeError:
        return None
    props = payload.get("data", {}).get("properties", {})
    declared: set[str] = set()
    _collect_flags(props, declared)
    return declared


def command_names(binary: str) -> list[str]:
    """Top-level verbs, straight from the binary's own inventory."""
    proc = subprocess.run(
        [binary, "--json", "commands"],
        capture_output=True,
        text=True,
        check=False,
    )
    proc.check_returncode()
    data = json.loads(proc.stdout).get("data", {})
    raw = data.get("commands", [])
    names = []
    for item in raw:
        if isinstance(item, str):
            names.append(item)
        elif isinstance(item, dict) and "name" in item:
            names.append(item["name"])
    return names


def main() -> int:
    binary = sys.argv[1] if len(sys.argv) > 1 else "./target/release/browser-automation-cli"

    globals_ = flags_in(run(binary, ["--help"])) | {"--help", "--version"}

    missing: dict[str, set[str]] = {}
    no_schema: list[str] = []
    checked = 0

    for name in command_names(binary):
        top_help = run(binary, [name, "--help"])
        exposed = flags_in(top_help)
        for sub in sorted(subcommands_in(top_help)):
            exposed |= flags_in(run(binary, [name, sub, "--help"]))
        exposed -= globals_
        if not exposed:
            continue

        declared = schema_flags(binary, name)
        if declared is None:
            no_schema.append(name)
            continue

        checked += 1
        if gap := exposed - declared:
            missing[name] = gap

    for name in sorted(missing):
        for flag in sorted(missing[name]):
            print(f"FAIL  {name}  {flag}  (accepted by clap, absent from schema)")
    for name in sorted(no_schema):
        print(f"FAIL  {name}  (no schema published at all)")

    total = sum(len(v) for v in missing.values()) + len(no_schema)
    print("----")
    print(f"commands_with_flags={checked}  undocumented_flags={total}")

    if total:
        print("== clap-schema-parity FAILED ==")
        print("Add the flag to src/commands/meta/schema/, then regenerate:")
        print("  bash scripts/generate_command_schemas.sh")
        return 1
    print("== clap-schema-parity PASS ==")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: tests/clap_schema_parity.rs
// SPDX-License-Identifier: MIT OR Apache-2.0
//! Clap-vs-schema parity: every flag clap accepts must appear in `schema --cmd`.
//!
//! # Why this exists as a second gate
//!
//! `scripts/schema-drift-check.sh` compares `docs/schemas/*.json` against the
//! live binary and is correct on its own axis. It cannot see this failure,
//! because both sides of that comparison derive from the SAME hand-written
//! schema module: the generator asks the binary, the binary asks
//! `src/commands/meta/schema/*.rs`. When a flag is added to the clap enum and
//! wired in the dispatcher, nothing in that loop forces a schema edit, so the
//! file and the binary agree perfectly about a surface neither describes.
//!
//! An agent discovers what it may pass by reading `schema`. A flag absent there
//! does not exist for the agent, whatever the parser happens to accept.
//!
//! # Why this is a Rust test and not a script
//!
//! It was `scripts/clap_schema_parity.py` until 2026-08-18. The product is Rust
//! end to end and ships no interpreter, so the gate died on any host without
//! that interpreter — and it is reached from `scripts/ci-check.sh`, the product's own
//! closure criterion. The port also buys the correct binary path:
//! `CARGO_BIN_EXE_` is resolved by cargo at compile time, where the script had
//! to guard by hand against measuring a stale artifact.
//!
//! # What this gate first reported, and why it was wrong
//!
//! The first cut reported 29 undocumented flags across `console`, `heap`,
//! `mitm`, `net`, `perf` and `storage`. Every one was a false positive with a
//! single cause: the clap side unions `<cmd> --help` with `<cmd> <sub> --help`,
//! so it sees two levels, while the schema side read one. A command that
//! dispatches on an action word keeps each action's arguments under
//! `action.actions.<name>.properties`. See [`collect_flags`], which recurses.
//!
//! Recorded rather than deleted, because the failure mode matters more than the
//! finding: a gate is a measuring instrument, and an instrument that disagrees
//! with the artifact is not evidence about the artifact until the instrument
//! itself has been checked.
//!
//! # Why `--help` and not a parse of the Rust source
//!
//! The same reason the drift gate trusts the binary: `--help` is what clap
//! actually built, so it cannot disagree with the parser. A regex over
//! `src/cli/` would introduce a third opinion, and a third opinion is a third
//! thing that can be wrong on its own.

use std::collections::{BTreeMap, BTreeSet};

use serde_json::Value;

mod common;

/// A near-empty universe means the walk failed, and every command would then
/// look clean. Measured 2026-08-28: 71 commands in the live inventory.
const MIN_COMMANDS: usize = 40;

/// Capture stdout plus stderr of a help invocation; clap may use either.
fn run(args: &[&str]) -> String {
    match common::cmd().args(args).output() {
        Ok(o) => format!(
            "{}{}",
            String::from_utf8_lossy(&o.stdout),
            String::from_utf8_lossy(&o.stderr)
        ),
        Err(_) => String::new(),
    }
}

/// Every long flag clap prints for this invocation.
///
/// Mirrors `^\s{2,}(?:-[a-zA-Z], )?(--[a-z0-9][a-z0-9-]*)`: an indented line
/// whose first token is a long flag, optionally preceded by its short alias.
fn flags_in(text: &str) -> BTreeSet<String> {
    let mut out = BTreeSet::new();
    for line in text.lines() {
        if !line.starts_with("  ") {
            continue;
        }
        let trimmed = line.trim_start();
        // `-q, --quiet` — step over the short alias and its separator.
        let candidate = if trimmed.len() > 3
            && trimmed.starts_with('-')
            && !trimmed.starts_with("--")
            && trimmed.as_bytes()[1].is_ascii_alphanumeric()
            && trimmed.as_bytes()[2] == b','
            && trimmed.as_bytes()[3] == b' '
        {
            &trimmed[4..]
        } else {
            trimmed
        };
        let Some(name) = long_flag_at_start(candidate) else {
            continue;
        };
        out.insert(name);
    }
    out
}

/// `--flag[=<V>]`, `--flag <V>` and bare `--flag` all yield `--flag`.
fn long_flag_at_start(s: &str) -> Option<String> {
    let b = s.as_bytes();
    if b.len() < 3 || b[0] != b'-' || b[1] != b'-' {
        return None;
    }
    if !(b[2].is_ascii_lowercase() || b[2].is_ascii_digit()) {
        return None;
    }
    let mut end = 3;
    while end < b.len()
        && (b[end].is_ascii_lowercase() || b[end].is_ascii_digit() || b[end] == b'-')
    {
        end += 1;
    }
    Some(s[..end].to_string())
}

/// Names listed under the `Commands:` block, if the command has one.
///
/// Mirrors `^\s{2}([a-z][a-z0-9-]*)\s{2,}\S`.
fn subcommands_in(text: &str) -> BTreeSet<String> {
    let mut out = BTreeSet::new();
    let mut in_block = false;
    for line in text.lines() {
        if line.starts_with("Commands:") {
            in_block = true;
            continue;
        }
        if !in_block {
            continue;
        }
        // A non-indented line ends the block (`Options:`, `Arguments:`, ...).
        if !line.is_empty() && !line.starts_with(' ') {
            break;
        }
        let Some(rest) = line.strip_prefix("  ") else {
            continue;
        };
        if rest.starts_with(' ') {
            continue;
        }
        let name: String = rest
            .chars()
            .take_while(|c| c.is_ascii_lowercase() || c.is_ascii_digit() || *c == '-')
            .collect();
        if name.is_empty() || name == "help" || !name.starts_with(|c: char| c.is_ascii_lowercase())
        {
            continue;
        }
        let after = &rest[name.len()..];
        if after.starts_with("  ") && !after.trim_start().is_empty() {
            out.insert(name);
        }
    }
    out
}

/// Add every long flag a properties object declares, at ANY nesting depth.
///
/// # Why the walk recurses
///
/// A command that dispatches on an action word keeps each action's arguments
/// under `action.actions.<name>.properties`, not at the top level. Reading only
/// the top level while the clap side already unions `<cmd> <sub> --help`
/// compares two different depths, and every nested flag is then reported as
/// undocumented — 29 findings of fictional debt, measured 2026-08-06.
///
/// # Why the key is a fallback for `argv`
///
/// Only some properties carry an explicit `argv`; most identify themselves by
/// key alone. A first cut that collected `argv` and nothing else reported 127
/// undocumented flags, including `qr --text`, which the schema documents as the
/// property `text`. The property key IS the documentation when `argv` is
/// absent, so it counts.
fn collect_flags(props: &serde_json::Map<String, Value>, out: &mut BTreeSet<String>) {
    for (key, prop) in props {
        let key_flag = format!("--{}", key.replace('_', "-"));
        let Some(obj) = prop.as_object() else {
            out.insert(key_flag);
            continue;
        };
        match obj.get("argv").and_then(Value::as_str) {
            Some(argv) if argv.starts_with("--") => {
                out.insert(argv.to_string());
            }
            _ => {
                out.insert(key_flag);
            }
        }
        if let Some(actions) = obj.get("actions").and_then(Value::as_object) {
            for spec in actions.values() {
                if let Some(nested) = spec.get("properties").and_then(Value::as_object) {
                    collect_flags(nested, out);
                }
            }
        }
    }
}

/// Long flags the published schema declares for `command`.
///
/// `None` means the command has NO schema at all, which is a different failure
/// from a schema that omits flags and is reported separately.
/// Flags the published schema declares for `command`, or `None` when the
/// product publishes no schema for it.
///
/// # Why every other failure panics instead of returning `None`
///
/// This function used to answer `None` for three unrelated events: the binary
/// failed to spawn, the binary exited non-zero for any reason, and the binary
/// answered that the command has no schema. The caller records `None` as
/// "no schema published at all" and the assertion tells a human to go write one.
///
/// On 2026-08-25 that misreport fired for real. Under load — a full pre-publish
/// gate compiling in parallel — one spawn of `schema --cmd extract` came back
/// unsuccessful, and the gate reported `FAIL extract (no schema published at
/// all)` and failed the release. Measured immediately afterwards, the schema
/// was there: exit 0 and 4118 bytes of JSON. Three consecutive re-runs passed.
///
/// So the exit code is read instead of thrown away. `2` is the product's usage
/// exit and is the only one that means "no such schema"; a spawn error, a
/// signal, or unparseable stdout on a successful exit are infrastructure
/// failures, and they now fail loudly and say what happened. A gate that cannot
/// tell "I could not look" from "it is not there" reports the second and sends
/// someone hunting for a file that exists.
fn schema_flags(command: &str) -> Option<BTreeSet<String>> {
    let out = common::cmd()
        .args(["--json", "schema", "--cmd", command])
        .output()
        .unwrap_or_else(|e| panic!("could not spawn the binary for `schema --cmd {command}`: {e}"));
    match out.status.code() {
        Some(0) => {}
        // The product's usage exit: this command genuinely publishes no schema.
        Some(2) => return None,
        other => panic!(
            "`schema --cmd {command}` exited {other:?}, which is neither success \
             nor the usage exit that means 'no schema'. This is a failure to \
             MEASURE, not a missing schema; do not add a schema on account of \
             it.\nstderr: {}",
            String::from_utf8_lossy(&out.stderr)
        ),
    }
    let payload: Value = serde_json::from_slice(&out.stdout).unwrap_or_else(|e| {
        panic!(
            "`schema --cmd {command}` exited 0 but stdout is not JSON: {e}\n\
             stdout: {}",
            String::from_utf8_lossy(&out.stdout)
        )
    });
    let mut declared = BTreeSet::new();
    if let Some(props) = payload
        .get("data")
        .and_then(|d| d.get("properties"))
        .and_then(Value::as_object)
    {
        collect_flags(props, &mut declared);
    }
    Some(declared)
}

/// Top-level verbs, straight from the binary's own inventory.
fn command_names() -> Vec<String> {
    let out = common::cmd()
        .args(["--json", "commands"])
        .output()
        .expect("commands --json must run");
    assert!(out.status.success(), "commands --json exited non-zero");
    let payload: Value = serde_json::from_slice(&out.stdout).expect("commands --json must be JSON");
    let items = payload
        .get("data")
        .and_then(|d| d.get("commands"))
        .and_then(Value::as_array)
        .expect("commands --json missing data.commands");
    items
        .iter()
        .filter_map(|item| match item {
            Value::String(s) => Some(s.clone()),
            Value::Object(o) => o.get("name").and_then(Value::as_str).map(str::to_string),
            _ => None,
        })
        .collect()
}

#[test]
fn every_clap_flag_appears_in_schema() {
    let mut globals = flags_in(&run(&["--help"]));
    globals.insert("--help".to_string());
    globals.insert("--version".to_string());

    let names = command_names();
    assert!(
        names.len() >= MIN_COMMANDS,
        "inventory collapsed to {} commands; the walk failed rather than the product shrinking",
        names.len()
    );

    let mut missing: BTreeMap<String, BTreeSet<String>> = BTreeMap::new();
    let mut no_schema: Vec<String> = Vec::new();
    let mut checked = 0usize;

    for name in &names {
        let top_help = run(&[name, "--help"]);
        let mut exposed = flags_in(&top_help);
        for sub in subcommands_in(&top_help) {
            exposed.extend(flags_in(&run(&[name, &sub, "--help"])));
        }
        for g in &globals {
            exposed.remove(g);
        }
        if exposed.is_empty() {
            continue;
        }

        let Some(declared) = schema_flags(name) else {
            no_schema.push(name.clone());
            continue;
        };

        checked += 1;
        let gap: BTreeSet<String> = exposed.difference(&declared).cloned().collect();
        if !gap.is_empty() {
            missing.insert(name.clone(), gap);
        }
    }

    let mut report = String::new();
    for (name, flags) in &missing {
        for flag in flags {
            report.push_str(&format!(
                "FAIL  {name}  {flag}  (accepted by clap, absent from schema)\n"
            ));
        }
    }
    for name in &no_schema {
        report.push_str(&format!("FAIL  {name}  (no schema published at all)\n"));
    }

    let total: usize = missing.values().map(BTreeSet::len).sum::<usize>() + no_schema.len();
    println!("----");
    println!("commands_with_flags={checked}  undocumented_flags={total}");

    assert!(
        total == 0,
        "{report}== clap-schema-parity FAILED ==\n\
         Add the flag to src/commands/meta/schema/, then regenerate:\n  \
         bash scripts/generate_command_schemas.sh"
    );
    println!("== clap-schema-parity PASS ==");
}
